File: Maximum Subarray.py
'''
https://leetcode.com/problems/maximum-subarray

Solved using Kadane\'s Algorithm
https://www.youtube.com/watch?v=86CQq3pKSUw
'''
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxSum = nums[0]
        thisSum = nums[0]
        for i in range(1, len(nums)):
            if(nums[i] > thisSum+nums[i]):
                thisSum = nums[i]
            else:
                thisSum += nums[i]
            if(thisSum > maxSum):
                maxSum = thisSum
        return(maxSum)

File: MaximumSubarray.c
/*

LeetCode 53. Maximum Subarray
https://leetcode.com/problems/maximum-subarray/

*/

int maxSubArray(int* nums, int numsSize){
    int max = nums[0];
    int max_now = nums[0];
    
    for(int i=1; i<numsSize; i++){
        if( max_now+nums[i] < nums[i] )
            max_now = nums[i];
        else
            max_now += nums[i];
        
        if( max_now > max )
            max = max_now;
        
    }
    
    return max;   
}